### PR TITLE
#763 Implement Segment

### PR DESCRIPTION
**Pull request by stephanevieira75**
# Preview

https://user-images.githubusercontent.com/36481167/105204963-fe764680-5b44-11eb-97c6-ad2e9bd7e328.mp4

### CODE REVIEW

File: twake/backend/node/package.json [hunk 1]
```diff
@@ -76,8 +76,10 @@
     "typescript": "^4.0.3"
   },
   "dependencies": {
+    "@types/analytics-node": "^3.1.4",
```
Comment (chamerling) [inline twake/backend/node/package.json:79]: This goes in `devDependencies`

File: twake/backend/node/src/core/platform/services/tracker/index.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    if ((await this.getAnalytics("segment")) && identity)
+      return (await this.getAnalytics("segment")).identify(identity, callback);
+  }
+
+  public async track(
+    tracker: TrackedEventType,
+    callback?: (err: Error) => void,
+  ): Promise<Analytics> {
+    if ((await this.getAnalytics("segment")) && tracker)
+      return (await this.getAnalytics("segment")).track(tracker, callback);
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/tracker/index.ts:24]: You await the same thing too many times, in this function and above. Please refactor

File: twake/backend/node/src/core/platform/services/tracker/provider.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import Analytics from "analytics-node";
+import { TwakeServiceProvider } from "../../framework";
+
+export type IdentityType =
+  | { userId: string | number }
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/tracker/provider.ts:5]: Do you need the first `|` ?

**Review by chamerling — CHANGES_REQUESTED** [review]
Adding the tracker everywhere we need to track pollutes the code. I suggest another approach: Use a local pubsub component.

The local pubsub component is 'local' to the backend instance. It allows to publish and subscribe to topics like we do for the AMQP one:

- On each event that we think important enough, we publish it
- When activated, the tracker component subscribe to all the events which are useful to it, and then push them to segment or elsewhere.

By doing this, we do not pollute code with tracker, we are also able to add more subscribers which are not tracker ones but are interested in the events.

This is quite simple to implement, check what I did for distributed pubsub, you can do the same (cf twake/backend/node/src/core/platform/services/pubsub/bus.ts) but create this dedicated local event bus in a new component, it can be a single instance in twake/backend/node/src/core/platform/framework/bus.ts

File: twake/backend/node/src/core/platform/services/tracker/provider.ts (hunk 3 above)
Comment (RomaricMourgues) [reply twake/backend/node/src/core/platform/services/tracker/provider.ts:5]: This one is probably code formater, if yes you can let it like this @stephanevieira75

**Comment by RomaricMourgues** [comment]
> Adding the tracker everywhere we need to track pollutes the code. I suggest another approach: Use a local pubsub component.
> 
> The local pubsub component is 'local' to the backend instance. It allows to publish and subscribe to topics like we do for the AMQP one:
> 
> * On each event that we think important enough, we publish it
> * When activated, the tracker component subscribe to all the events which are useful to it, and then push them to segment or elsewhere.
> 
> By doing this, we do not pollute code with tracker, we are also able to add more subscribers which are not tracker ones but are interested in the events.
> 
> This is quite simple to implement, check what I did for distributed pubsub, you can do the same (cf twake/backend/node/src/core/platform/services/pubsub/bus.ts) but create this dedicated local event bus in a new component, it can be a single instance in twake/backend/node/src/core/platform/framework/bus.ts

Just to be sure, this doesn't all rabbitmq ? I don't want to add rabbitmq calls for the tracker :)

**Comment by chamerling** [comment]
@RomaricMourgues no it is not rabbit. What I call 'local' pubsub is an EventEmitter instance at the end

**Comment by RomaricMourgues** [comment]
> @RomaricMourgues no it is not rabbit. What I call 'local' pubsub is an EventEmitter instance at the end

Perfect then !

File: twake/backend/node/package.json (hunk 1 above)
Comment (stephanevieira75) [reply twake/backend/node/package.json]: Done :)

File: twake/backend/node/src/core/platform/services/tracker/index.ts (hunk 2 above)
Comment (stephanevieira75) [reply twake/backend/node/src/core/platform/services/tracker/index.ts]: done 👍🏻

File: twake/backend/node/src/core/platform/services/tracker/provider.ts (hunk 3 above)
Comment (stephanevieira75) [reply twake/backend/node/src/core/platform/services/tracker/provider.ts]: Yes, it is code formater !

File: twake/backend/node/config/default.json [hunk 4]
```diff
@@ -9,11 +9,15 @@
...
-    "phpnode"
+    "phpnode",
+    "tracker"
   ],
   "logger": {
     "level": "debug"
   },
+  "tracker": {
+    "segment": ""
+  },
```
Comment (chamerling) [inline twake/backend/node/config/default.json:20]: If we want multiple trackers (or to follow the same format as others), it should be:

```js
"tracker":
  "type": "segment",
  "segment": {
    "secret": "the secret"
  }
```

File: twake/backend/node/src/core/platform/services/tracker/index.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+    return this;
+  }
+
+  private getVisibilityObject(visibility: string) {
+    return {
+      is_direct: visibility === "direct" ? true : false,
+      is_private: visibility === "private" ? true : false,
+      is_public: visibility === "public" ? true : false,
+    };
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/tracker/index.ts:85]: Simple way to do:

```js
is_direct: !!(visibility === "direct)
```

File: twake/backend/node/src/core/platform/services/tracker/index.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      this.identify({
+        userId: data.user.id,
+      });
+      this.track(
+        {
+          userId: data.user.id,
+          event: TrackerEventActions.TWAKE_OPEN_CLIENT,
+        },
+        (err: Error) => (err ? logger.error("Error while tracking -->", err) : false),
+      );
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/tracker/index.ts:29]: The `identity` and `track` functions are async, do you want to call identity then track or both in any order?

File: twake/backend/node/src/core/platform/services/tracker/index.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    if (analytics && tracker) return analytics.track(tracker, callback);
+  }
+
+  private async getAnalytics(trackerName: string): Promise<Analytics> {
+    const trackerKey = this.configuration.get<string>(trackerName, "");
+
+    if (!this.analytics && trackerKey) {
+      this.analytics = new Analytics(trackerKey);
+    }
+    return this.analytics;
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/tracker/index.ts:112]: You should add a comment in the function saying that it may return undefined. Also, you must change the function return type because this.analytics can be undefined

File: twake/backend/node/src/services/channels/index.ts [hunk 8]
```diff
@@ -7,7 +7,7 @@ import { DatabaseServiceAPI } from "../../core/platform/services/database/api";
 import { PubsubServiceAPI } from "../../core/platform/services/pubsub/api";
 
 @Prefix("/internal/services/channels/v1")
-@Consumes(["webserver", "database", "pubsub"])
+@Consumes(["webserver", "database", "pubsub", "tracker"])
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/index.ts:10]: As we said, this should not be required here

File: twake/backend/node/src/services/notifications/index.ts [hunk 9]
```diff
@@ -7,7 +7,7 @@ import { DatabaseServiceAPI } from "../../core/platform/services/database/api";
 import { PubsubServiceAPI } from "../../core/platform/services/pubsub/api";
 
 @Prefix("/internal/services/notifications/v1")
-@Consumes(["webserver", "database", "pubsub"])
+@Consumes(["webserver", "database", "pubsub", "tracker"])
```
Comment (chamerling) [inline twake/backend/node/src/services/notifications/index.ts:10]: Also should not be required

File: twake/backend/node/src/services/notifications/services/engine/processors/new-channel-message/index.ts [hunk 10]
```diff
@@ -9,6 +9,11 @@ import {
 } from "../../../../entities";
 import { ChannelMemberNotificationLevel } from "../../../../../channels/types";
 import { MentionNotification } from "../../../../types";
+import { trackedEventBus } from "../../../../../../core/platform/framework/pubsub";
```
Comment (chamerling) [inline twake/backend/node/src/services/notifications/services/engine/processors/new-channel-message/index.ts:12]: `trackedEventBus` is not a good name. Name it like `localBus` (same for other files)

File: twake/backend/node/src/core/platform/framework/pubsub.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
+import { EventEmitter } from "events";
+import { TrackerEventActions } from "../../../core/platform/services/tracker/types";
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/framework/pubsub.ts:2]: This is not related to tracker, this event bus is independant. The type should also be local to the pubsub.ts file (put it in this file)

File: twake/backend/node/src/core/platform/framework/pubsub.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  constructor() {
+    super();
+  }
+
+  /**
+   * Subscribe is MUST only be used by the pubsub service
+   *
+   * @param listener
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/framework/pubsub.ts:20]: I do not see any `any` :)

File: twake/backend/node/src/core/platform/framework/pubsub.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  subscribe<T>(name: TrackerEventActions, listener: (data: T) => void): this {
+    return this.on(name, listener);
+  }
+
+  /**
+   * Publish must only be used by the Pubsub decorators
+   *
+   * @param data
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/framework/pubsub.ts:30]: Same here, no `any`, so remove this comment

File: twake/backend/node/src/core/platform/framework/pubsub.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  publish<T>(name: TrackerEventActions, data: T): boolean {
+    return this.emit(name, data);
+  }
+}
+
+export const trackedEventBus = new TrackedEventBus();
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/framework/pubsub.ts:36]: `export const localEventBus = new EventBus();`

(and rename the class to `EventBus`)
